File: apps/system/api/interface/presenter/workspace/invitation/adapter.go
package invitation

import (
	"github.com/ryo034/react-go-template/apps/system/api/domain/workspace/invitation"
	"github.com/ryo034/react-go-template/apps/system/api/interface/presenter/member"
	"github.com/ryo034/react-go-template/apps/system/api/schema/openapi"
)

type Adapter interface {
	Adapt(i *invitation.Invitation) (openapi.Invitation, error)
	AdaptAll(is invitation.Invitations) (openapi.Invitations, error)
}

type adapter struct {
	ma member.Adapter
}

func NewAdapter(ma member.Adapter) Adapter {
	return &adapter{ma}
}

func (a *adapter) Adapt(i *invitation.Invitation) (openapi.Invitation, error) {
	if i == nil {
		return openapi.Invitation{}, nil
	}
	lt, err := i.ExpiredAt().Value().ToLocalTime()
	if err != nil {
		return openapi.Invitation{}, err
	}
	dn := ""
	if i.DisplayName() != nil {
		dn = i.DisplayName().ToString()
	}
	return openapi.Invitation{
		ID:           i.ID().Value(),
		Accepted:     i.IsAccepted(),
		ExpiredAt:    lt,
		InviteeEmail: i.InviteeEmail().ToString(),
		DisplayName:  dn,
		Inviter:      a.ma.Adapt(i.Inviter()),
	}, nil
}

func (a *adapter) AdaptAll(is invitation.Invitations) (openapi.Invitations, error) {
	if is == nil {
		return nil, nil
	}
	res := make([]openapi.Invitation, 0, is.Size())
	for _, w := range is.AsSlice() {
		a, err := a.Adapt(w)
		if err != nil {
			return nil, err
		}
		res = append(res, a)
	}
	return res, nil
}

File: apps/system/client/src/interface/gateway/workspace/invitation/adapter.ts
import { Result } from "true-myth"
import { Invitation, InvitationId, Invitations, MemberDisplayName } from "~/domain"
import { AppDateTime, Email } from "~/domain/shared"
import { components } from "~/generated/schema/openapi/systemApi"
import { AdapterError } from "~/infrastructure/error"
import { MemberGatewayAdapter } from "../member"

export class InvitationGatewayAdapter {
  constructor(private readonly memberAdapter: MemberGatewayAdapter) {}
  adapt(invitation: components["schemas"]["Invitation"]): Result<Invitation, Error> {
    if (invitation === undefined || invitation === null) {
      console.error(new AdapterError(InvitationGatewayAdapter.name, this.adapt.name, "Invitation is required"))
      return Result.err(new Error("Invitation is required"))
    }

    const id = InvitationId.fromString(invitation.id)
    if (id.isErr) {
      return Result.err(id.error)
    }

    let displayName: MemberDisplayName | null = null
    if (invitation.displayName !== null && invitation.displayName.length > 0) {
      const displayNameRes = MemberDisplayName.create(invitation.displayName)
      if (displayNameRes.isErr) {
        return Result.err(displayNameRes.error)
      }
      displayName = displayNameRes.value
    }

    const inviteeEmail = Email.create(invitation.inviteeEmail)
    if (inviteeEmail.isErr) {
      return Result.err(inviteeEmail.error)
    }

    const expiredAt = AppDateTime.fromString(invitation.expiredAt)
    if (expiredAt.isErr) {
      return Result.err(expiredAt.error)
    }

    const inviter = this.memberAdapter.adapt(invitation.inviter)
    if (inviter.isErr) {
      return Result.err(inviter.error)
    }

    return Result.ok(
      Invitation.create({
        id: id.value,
        displayName,
        inviteeEmail: inviteeEmail.value,
        accepted: invitation.accepted,
        expiredAt: expiredAt.value,
        inviter: inviter.value
      })
    )
  }

  adaptAll(invitations: components["schemas"]["Invitations"]): Result<Invitations, Error> {
    const vs: Invitation[] = []
    for (const i of invitations) {
      const res = this.adapt(i)
      if (res.isErr) {
        return Result.err(res.error)
      }
      vs.push(res.value)
    }
    return Result.ok(Invitations.create(vs))
  }
}
